cpp/include/cudf/detail/iterator.cuh

```diff
@@ -167,6 +167,86 @@ auto make_null_replacement_iterator(column_device_view const& column,
     0, null_replaced_value_accessor<Element>{column, null_replacement, has_nulls});
 }
 
+/**
+ * @brief Constructs an optional iterator over a column's values and its validity.
+ *
+ * Dereferencing the returned iterator returns a `thrust::optional<Element>`.
+ *
+ * When the element of an iterator contextually converted to bool, the conversion returns true
+ * if the object contains a value and false if it does not contain a value.
+ *
+ * make_optional_iterator with mode `DYNAMIC` defers the assumption of nullability to
+ * runtime, with the user stating on construction of the iterator if column has nulls.
+ *
+ * @throws cudf::logic_error if the column is not nullable, and `DYNAMIC` mode used and
+ *         the user has stated nulls exist
+ * @throws cudf::logic_error if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @tparam MODE The null_mode of the iterator
+ * @param column The column to iterate
+ * @return auto Iterator that returns valid column elements, and validity of the
+ * element in a thrust::optional
+ */
+template <typename Element>
+auto make_optional_iterator(column_device_view const& column,
+                            contains_nulls::DYNAMIC mode,
+                            bool has_nulls = true)
+{
+  return column.optional_begin<Element>(mode, has_nulls);
+}
+
+/**
+ * @brief Constructs an optional iterator over a column's values and its validity.
+ *
+ * Dereferencing the returned iterator returns a `thrust::optional<Element>`.
+ *
+ * When the element of an iterator contextually converted to bool, the conversion returns true
+ * if the object contains a value and false if it does not contain a value.
+ *
+ * make_optional_iterator ith mode `YES` means that the column supports nulls and
+ * potentially has null values, therefore the optional might not contain a value
+ *
+ * @throws cudf::logic_error if the column is not nullable, and `YES` mode used
+ * @throws cudf::logic_error if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @tparam mode The has_nulls mode of the iterator
+ * @param column The column to iterate
+ * @return auto Iterator that returns column elements, and validity of the
+ * element in a thrust::optional
+ */
+template <typename Element>
+auto make_optional_iterator(column_device_view const& column, contains_nulls::YES mode)
+{
+  return column.optional_begin<Element>(mode);
+}
```

Comment:
I think having the param unnamed makes it clearer that it's a tag type. 
```suggestion
auto make_optional_iterator(column_device_view const& column, contains_nulls::YES)
{
  return column.optional_begin<Element>(contains_nulls::YES{});
}
```

Reply:
I can do that. I used a name to reduce the chances of a copy and paste error when stamping out the three versions.

Reply:
I see your point. I just worry people will get confused and think `mode` is useful for anything other than it's type.